From: IrfanMohammad <irfan@graphiant.com>
Date: Fri, 9 Aug 2024 12:15:24 +0000
Subject: always attempt recovery on delayed rwlock acquire

With every retry attempt of acquiring a rwlock, we can also check the
aliveness of the CID. A dead CID has the ability to cause system-wide
delays until the lock is recovered.

This mainly entails doing a CID liveness check on all lock holders.
The current mechanism for liveness check involves checking if a file is
locked by the owning process. A faster and better mechanism can be
implemented but is not trivial and from profiling for the performance
impact, the current mechanism only incurs a max penalty of < 10 us each
iteration.
---
 src/common.c  | 8 ++++++++
 src/shm_sub.c | 1 +
 2 files changed, 9 insertions(+)

diff --git a/src/common.c b/src/common.c
index c5d4242..ef0041f 100644
--- a/src/common.c
+++ b/src/common.c
@@ -65,6 +65,8 @@
 
 #define SR_IS_YANG_ID_CHAR(c) (isalpha(c) || isdigit(c) || ((c) == '_') || ((c) == '-') || ((c) == '.'))
 
+#define SR_COND_RETRY_STEP_SEC 3
+
 #define SR_RWLOCK_ID(rwlock_ptr) (~(1u<<31) & (rwlock_ptr)->id)
 #define _SR_LOG_RWLOCK(ll, rwlock_ptr, mode, cid, func, str)           \
     sr_log(ll, "%s: %s rwlock %" PRIu32 " in mode %d by CID %" PRIu32, \
@@ -2673,6 +2675,7 @@ sr_sub_rwlock(sr_rwlock_t *rwlock, struct timespec *timeout_abs, sr_lock_mode_t
             ret = sr_cond_clockwait_retry(&rwlock->cond, &rwlock->mutex, timeout_abs);
             if (ret == EAGAIN) {
                 SR_RWLOCK_WRN(rwlock, mode, cid, func, "delay locking");
+                sr_rwlock_recover(rwlock, func, cb, cb_data);
                 ret = 0;
             }
         }
@@ -2715,6 +2718,7 @@ sr_sub_rwlock(sr_rwlock_t *rwlock, struct timespec *timeout_abs, sr_lock_mode_t
             ret = sr_cond_clockwait_retry(&rwlock->cond, &rwlock->mutex, timeout_abs);
             if (ret == EAGAIN) {
                 SR_RWLOCK_WRN(rwlock, mode, cid, func, "delay locking");
+                sr_rwlock_recover(rwlock, func, cb, cb_data);
                 ret = 0;
             }
         }
@@ -2757,6 +2761,7 @@ sr_sub_rwlock(sr_rwlock_t *rwlock, struct timespec *timeout_abs, sr_lock_mode_t
             ret = sr_cond_clockwait_retry(&rwlock->cond, &rwlock->mutex, timeout_abs);
             if (ret == EAGAIN) {
                 SR_RWLOCK_WRN(rwlock, mode, cid, func, "delay locking");
+                sr_rwlock_recover(rwlock, func, cb, cb_data);
                 ret = 0;
             }
         }
@@ -2798,6 +2803,7 @@ sr_sub_rwlock(sr_rwlock_t *rwlock, struct timespec *timeout_abs, sr_lock_mode_t
             ret = sr_cond_clockwait_retry(&rwlock->cond, &rwlock->mutex, timeout_abs);
             if (ret == EAGAIN) {
                 SR_RWLOCK_WRN(rwlock, mode, cid, func, "delay locking");
+                sr_rwlock_recover(rwlock, func, cb, cb_data);
                 ret = 0;
             }
         }
@@ -2928,6 +2934,7 @@ sr_rwrelock(sr_rwlock_t *rwlock, uint32_t timeout_ms, sr_lock_mode_t mode, sr_ci
             ret = sr_cond_clockwait_retry(&rwlock->cond, &rwlock->mutex, &timeout_abs);
             if (ret == EAGAIN) {
                 SR_RWLOCK_WRN(rwlock, mode, cid, func, "delay relocking");
+                sr_rwlock_recover(rwlock, func, cb, cb_data);
                 ret = 0;
             }
         }
@@ -2994,6 +3001,7 @@ sr_rwrelock(sr_rwlock_t *rwlock, uint32_t timeout_ms, sr_lock_mode_t mode, sr_ci
             ret = sr_cond_clockwait_retry(&rwlock->cond, &rwlock->mutex, &timeout_abs);
             if (ret == EAGAIN) {
                 SR_RWLOCK_WRN(rwlock, mode, cid, func, "delay relocking");
+                sr_rwlock_recover(rwlock, func, cb, cb_data);
                 ret = 0;
             }
         }
diff --git a/src/shm_sub.c b/src/shm_sub.c
index 8c9d99f..f578f92 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -381,6 +381,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
             snprintf(buf, sizeof(buf), "orig-CID %" PRIu32 " request ID %" PRIuFAST32,
                     sub_shm->orig_cid, ATOMIC_LOAD_RELAXED(sub_shm->request_id));
             SR_RWLOCK_WRN(&sub_shm->lock, SR_LOCK_NONE, cid, "delay new-wrlock", buf);
+            sr_shmsub_recover(sub_shm);
             ret = 0;
         }
     }
